src/osdc/Objecter.h

```diff
@@ -1250,6 +1250,55 @@ class Objecter : public md_config_obs_t, public Dispatcher {
   // to be drained by consume_blacklist_events.
   bool blacklist_events_enabled;
   std::set<entity_addr_t> blacklist_events;
+  struct pg_mapping_t {
+    std::vector<int> up;
+    int up_primary = -1;
+    std::vector<int> acting;
+    int acting_primary = -1;
+
+    pg_mapping_t() {}
+    pg_mapping_t(std::vector<int> up, int up_primary,
+                 std::vector<int> acting, int acting_primary)
+               : up(up), up_primary(up_primary),
+                 acting(acting), acting_primary(acting_primary) {}
+  };
+  std::shared_mutex pg_mapping_lock;
+  std::map<int64_t, std::map<pg_t, std::pair<epoch_t, pg_mapping_t>>> pg_mappings;
```

Comment:
I was expecting something like ``map<int64_t,vector<pg_mapping_t>>`` where pg_mapping_t has an ``epoch`` member to indicate whether it's valid/uptodate.  But if we're seeing speedup from a plain std::map then this is a perfectly reasonable step in that direction!